p11-kit/import-object.c

```diff
@@ -0,0 +1,606 @@
+ *       following disclaimer.
+ *     * Redistributions in binary form must reproduce the
+ *       above copyright notice, this list of conditions and
+ *       the following disclaimer in the documentation and/or
+ *       other materials provided with the distribution.
+ *     * The names of contributors to this software may not be
+ *       used to endorse or promote products derived from this
+ *       software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * Author: Zoltan Fridrich <zfridric@redhat.com>
+ */
+
+#include "config.h"
+
+#include "attrs.h"
+#include "compat.h"
+#include "debug.h"
+#include "dict.h"
+#include "iter.h"
+#include "message.h"
+#include "pem.h"
+#include "tool.h"
+
+#ifdef OS_UNIX
+#include "tty.h"
+#endif
+
+#ifdef WITH_ASN1
+#include "asn1.h"
+#include "oid.h"
+#endif
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef ENABLE_NLS
+#include <libintl.h>
+#define _(x) dgettext(PACKAGE_NAME, x)
+#else
+#define _(x) (x)
+#endif
+
+int
+p11_kit_import_object (int argc,
+		       char *argv[]);
+
+#ifdef WITH_ASN1
+
+typedef struct {
+	CK_FUNCTION_LIST *module;
+	CK_SESSION_HANDLE session;
+	const char *label;
+} import_data;
+
+static void
+import_x509_cert(const unsigned char *der,
+		 size_t der_len,
+		 const import_data *data)
+{
+	CK_RV rv;
+	int start = 0, end = 0;
+	p11_dict *defs = NULL;
+	asn1_node asn = NULL;
+	CK_OBJECT_HANDLE object = 0;
+	CK_ATTRIBUTE *attrs = NULL, *tmp = NULL;
+	CK_BBOOL tval = CK_TRUE, fval = CK_FALSE;
+	CK_OBJECT_CLASS class = CKO_CERTIFICATE;
+	CK_CERTIFICATE_TYPE cert_type = CKC_X_509;
+	CK_ATTRIBUTE attr_class = { CKA_CLASS, &class, sizeof (class) };
+	CK_ATTRIBUTE attr_token = { CKA_TOKEN, &tval, sizeof (tval) };
+	CK_ATTRIBUTE attr_private = { CKA_PRIVATE, &fval, sizeof (fval) };
+	CK_ATTRIBUTE attr_cert_type = { CKA_CERTIFICATE_TYPE, &cert_type, sizeof (cert_type) };
+	CK_ATTRIBUTE attr_value = { CKA_VALUE, (void *)der, der_len };
+	CK_ATTRIBUTE attr_subject = { CKA_SUBJECT, };
+
+	return_if_fail (data != NULL);
+	return_if_fail (data->module != NULL);
+	return_if_fail (data->session != 0);
+
+	defs = p11_asn1_defs_load ();
```

Comment:
Can this be moved to `import_pem` to avoid code duplication?

Reply:
yes